File: 2021/2021W20/PreppindataW20.py
import pandas as pd
from datetime import datetime as dt

out_file = "F:/Data/prepindata2021w20py.xlsx"

complaints = pd.read_csv("F:/Data/Prep Air Complaints - Complaints per Day.csv",
                         parse_dates=['Date'], date_parser=lambda x: dt.strptime(x, '%d/%m/%Y'))
complaints['Date'] = complaints['Date'].dt.date
complaints['Mean'] = complaints['Complaints'].groupby([complaints['Week']]).transform('mean')
complaints['Standard Deviation'] = complaints['Complaints'].groupby([complaints['Week']]).transform('std')

with pd.ExcelWriter(out_file) as writer:  
    for i in range(3):
        df = complaints.copy()
        df['Lower Control Limit'] = df['Mean'] - df['Standard Deviation']*(i+1)
        df['Upper Control Limit'] = df['Mean'] + df['Standard Deviation']*(i+1)
        df['Variation'] = df['Upper Control Limit'] - df['Lower Control Limit']
        df['Outlier'] = df.apply(lambda x: 'Inside' if x['Upper Control Limit']>=x['Complaints']>=x['Lower Control Limit'] else 'Outside', axis=1)
        df = df[df['Outlier']=='Outside']
        df.to_excel(writer, sheet_name=f'{i+1}SD', index=False)  

File: 2021/2021W20/preppindataW20.R
library(dplyr)
library(readr)
library(openxlsx)

Complaints <- read_csv("F:/Data/Prep Air Complaints - Complaints per Day.csv", 
                       col_types = cols(Date = col_date(format = "%d/%m/%Y"))) %>%
  group_by(`Week`) %>%
  mutate('Mean' = mean(`Complaints`),
         'Standard Deviation' = sd(`Complaints`))

NSD <- function(N){
  df <-Complaints %>%
    mutate('Lower Control Limit' = `Mean`-`Standard Deviation`*N,
           'Upper Control Limit' = `Mean`+`Standard Deviation`*N,
           'Variation' = `Upper Control Limit`-`Lower Control Limit`,
           'Outlier' = if_else(`Upper Control Limit`>=`Complaints` & `Complaints`>=`Lower Control Limit`,'Inside','Outside')) %>%
    filter(`Outlier`=='Outside')
  return(df)
}

fin_sheet <- list("1SD" = NSD(1), "2SD" = NSD(2), "3SD" = NSD(3))
write.xlsx(fin_sheet, file = "F:/Data/prepindata2021w20r.xlsx")
